<center>

$\Huge \textbf{Universidad Nacional Autónoma de México}$  
$\Huge \textbf{Facultad de Ciencias}$  
<p align="center">
  <img src="https://www.icat.unam.mx/wp-content/uploads/2021/11/Copia-de-LogoUNAM.-Azul.-Fondo-transparente.png" alt="UNAM" width="200"/>
</p>

<hr style="height:3px; background-color:#0B6E4F; border:none;"/>


$\LARGE \textbf{Inteligencia Artificial}$  

$\Large \textit{Laboratorio 5.8}$  


\begin{array}{rl}
\textbf{Docente:} & Dra. Jessica Sarahi Méndez Rincón \\[6pt]
\textbf{Ayudante de laboratorio:} & Diego Eduardo Peña Villegas \\[6pt]
\textbf{Alumna:} & Flores Perusquia Ricardo \\[6pt]
\textbf{Fecha de realización:} & 06/05/2026
\end{array}

</center>


**UNAM y la creadora del presente notebook, no guarda relación alguna con las marcas mencionadas como ejemplo. Las marcas son propiedad de sus titulares conforme a la legislación aplicable, se utilizan con fines académicos y didácticos, por lo que no existen fines de lucro, relación publicitaria o de patrocinio.

---

In [7]:
!pip install transformers

In [8]:
from transformers import pipeline, AutoTokenizer, AutoModelForSequenceClassification

# Cargar el modelo pre-entrenado y el tokenizador de BERT en español
#model_name = "dccuchile/bert-base-spanish-wwm-uncased"
# Este ya sabe qué es Positivo, Negativo y Neutral
# Muy robusto para español
model_name = "nlptown/bert-base-multilingual-uncased-sentiment"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name)

# Crear un pipeline para la clasificación de texto
text_classification_pipeline = pipeline(
    'text-classification',
    model=model,
    tokenizer=tokenizer
)

# Ejecutar la clasificación de texto en una oración en español
#result = text_classification_pipeline('Esta película es muy buena, la recomiendo totalmente')
#print(result)

# Ejecutar la clasificación
result = text_classification_pipeline('Esta película es muy buena, la recomiendo totalmente')

# Acceder a los datos dentro de la lista y el diccionario
label = result[0]['label']
score = result[0]['score']

print(f"Resultado: {label}")
print(f"Confianza: {score:.4f}")

Loading weights:   0%|          | 0/201 [00:00<?, ?it/s]

Resultado: 5 stars
Confianza: 0.7252


1. El cambio en las etiquetas (Labels)
Al usar un modelo ajustado (probablemente el de nlptown o uno similar de reseñas), el sistema ya no usa códigos genéricos como LABEL_1.

"5 stars": El modelo ha clasificado la frase ("Esta película es muy buena, la recomiendo totalmente") con la puntuación máxima. Ha detectado un sentimiento altamente positivo.

2. La Confianza (Score)
0.7252 (72.5%): Es una métrica sólida. Indica que, de todas las opciones posibles (1, 2, 3, 4 o 5 estrellas), el modelo está bastante seguro de que pertenece a la categoría de 5.

In [9]:
from transformers import AutoTokenizer, AutoModelForCausalLM

# Descargar el modelo pre-entrenado de GPT-2 en español
model_name = "mrm8488/spanish-gpt2"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

Loading weights:   0%|          | 0/148 [00:00<?, ?it/s]

GPT2LMHeadModel LOAD REPORT from: mrm8488/spanish-gpt2
Key                                     | Status     |  | 
----------------------------------------+------------+--+-
transformer.h.{0...11}.attn.bias        | UNEXPECTED |  | 
transformer.h.{0...11}.attn.masked_bias | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


In [10]:
#Cambiar este texto para 3 ejemplos
input_str = "Había una vez"
input_str = "El hombre bicentenario es"
input_str = "En un lugar muy muy lejano"

max_length = 100
# Generar texto con el modelo pre-entrenado de GPT-2 en español
input_ids = tokenizer.encode(input_str, return_tensors="pt")
sample_outputs = model.generate(
    input_ids,
    do_sample=True,
    max_length=max_length,
    top_k=50,
    top_p=0.95,
    temperature=1.0,
    num_return_sequences=1
)
generated_text = tokenizer.decode(sample_outputs[0], skip_special_tokens=True)



print(generated_text)

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


En un lugar muy muy lejano, un lugar como ningún otro, es muy peligroso.La mayoría de la gente dice:"Es realmente un viaje sin fin"."Es un lugar oscuro y profundo"."La vida aquí es muy intensa".Yo, yo no diría eso, ¿cierto?¿Dónde estás?¿En el bosque?No...- ¿Dónde estás?- Ahí.- ¿En el bosque?- Sí, en el bosque.¿Eres un cazador?Sí.¿Y qué


In [11]:
#Cambiar este texto para 3 ejemplos
input_str = "Había una vez"
input_str = "El hombre bicentenario es"
input_str = "En un lugar muy muy lejano"
max_length = 100
# Generar texto con el modelo pre-entrenado de GPT-2 en español
input_ids = tokenizer.encode(input_str, return_tensors="pt")
sample_outputs = model.generate(
    input_ids,
    do_sample=True,
    max_length=max_length,
    top_k=50,
    top_p=0.95,
    temperature=1.0,
    num_return_sequences=1
)
generated_text = tokenizer.decode(sample_outputs[0], skip_special_tokens=True)



print(generated_text)

The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


En un lugar muy muy lejano se encuentra el planeta Venus.Aquí está la Venus de 1815.Es un enorme cráter de 14 km de diámetro, en el centro del hemisferio norte.Está casi totalmente cubierto por el hielo.Lo llaman "la Tierra de Horus", pero es un planeta oscuro.Se ve como la superficie de Venus, pero en realidad es un planeta gigante y muy frío.Y, sin embargo, de todas las lunas de Saturno, ninguna es igual a la de Venus.La más


**Conclusion**

Basado en los dos últimos bloques del código, donde se generan textos a partir de tres frases iniciales diferentes ("Había una vez", "El hombre bicentenario es", "En un lugar muy muy lejano"), se concluye que el modelo es capaz de producir continuaciones coherentes y en español a partir de estímulos muy breves.

<hr/>
<footer style="text-align:center; font-size:12px; color:gray;">
© 2026 UNAM Facultado de Ciencias – Todos los derechos reservados

</footer>